druid-shell/src/mouse.rs

```diff
@@ -18,50 +18,218 @@ use crate::kurbo::Point;
     pub button: MouseButton,
 }
 
 /// An indicator of which mouse button was pressed.
 #[derive(PartialEq, Eq, Clone, Copy, Debug)]
+#[repr(u8)]
 pub enum MouseButton {
     /// Left mouse button.
     Left,
-    /// Middle mouse button.
-    Middle,
     /// Right mouse button.
     Right,
+    /// Middle mouse button.
+    Middle,
     /// First X button.
     X1,
     /// Second X button.
     X2,
+    /// No mouse button.
+    None,
 }
 
 impl MouseButton {
-    /// Returns `true` if this is the left mouse button.
-    #[inline(always)]
+    /// Returns `true` if this is [`MouseButton::Left`].
+    ///
+    /// [`MouseButton::Left`]: #variant.Left
+    #[inline]
     pub fn is_left(self) -> bool {
         self == MouseButton::Left
     }
 
-    /// Returns `true` if this is the right mouse button.
-    #[inline(always)]
+    /// Returns `true` if this is [`MouseButton::Right`].
+    ///
+    /// [`MouseButton::Right`]: #variant.Right
+    #[inline]
     pub fn is_right(self) -> bool {
         self == MouseButton::Right
     }
+
+    /// Returns `true` if this is [`MouseButton::Middle`].
+    ///
+    /// [`MouseButton::Middle`]: #variant.Middle
+    #[inline]
+    pub fn is_middle(self) -> bool {
+        self == MouseButton::Middle
+    }
+
+    /// Returns `true` if this is [`MouseButton::X1`].
+    ///
+    /// [`MouseButton::X1`]: #variant.X1
+    #[inline]
+    pub fn is_x1(self) -> bool {
+        self == MouseButton::X1
+    }
+
+    /// Returns `true` if this is [`MouseButton::X2`].
+    ///
+    /// [`MouseButton::X2`]: #variant.X2
+    #[inline]
+    pub fn is_x2(self) -> bool {
+        self == MouseButton::X2
+    }
+}
+
+/// A set of [`MouseButton`]s.
+///
+/// [`MouseButton`]: enum.MouseButton.html
+#[derive(PartialEq, Eq, Clone, Copy, Default)]
+pub struct MouseButtons(u8);
+
+impl MouseButtons {
+    /// Create a new empty set.
+    #[inline]
+    pub fn new() -> MouseButtons {
+        MouseButtons(0)
+    }
+
+    /// Add the `button` to the set.
+    #[inline]
+    pub fn add(&mut self, button: MouseButton) {
+        self.0 |= 1 << button as u8;
+    }
+
+    /// Remove the `button` from the set.
+    #[inline]
+    pub fn remove(&mut self, button: MouseButton) {
+        self.0 &= !(1 << button as u8);
+    }
+
+    /// Builder-style method for adding the `button` to the set.
+    #[inline]
+    pub fn with(mut self, button: MouseButton) -> MouseButtons {
+        self.0 |= 1 << button as u8;
+        self
+    }
+
+    /// Builder-style method for removing the `button` from the set.
+    #[inline]
+    pub fn without(mut self, button: MouseButton) -> MouseButtons {
+        self.0 &= !(1 << button as u8);
+        self
+    }
+
+    /// Returns `true` if the `button` is in the set.
+    #[inline]
+    pub fn has(self, button: MouseButton) -> bool {
+        (self.0 & (1 << button as u8)) != 0
+    }
+
+    /// Returns `true` if any button is in the set.
+    #[inline]
+    pub fn has_any(self) -> bool {
```

Comment:
I might prefer to use `is_empty()` for this, if we're treating it like a set? Not totally sure, just a thought.

Reply:
I named as many as I could with `has_` prefix because it seemed nicer to me. However it's probably more ergonomic to follow `HashSet` more closely in naming, because that's what Rust programmers will know. I will rename all the methods that don't match but have equals in `HashSet`.

Reply:
yea, this is my general thinking; people are more likely to try and use method names they know from elsewhere.